feat(core): add support for Types in ViewContainerRef.createComponent

AndrewKushnir opened the pull request:
Since main branch is now open for v13 changes, I'm reviving PR #41954 (as a new PR for better visibility).

With Ivy it's possible to create a `ComponentFactory` instance based on the generated Component def, thus the `ViewContainerRef.createComponent` can accept a Component instance and create a factory inside, rather than requiring additional complexity of creating a ComponentFactory outside (by invoking `ComponentFactoryResolver`).

This should simplify the API that is used for creating components dynamically.

Current version:

```
@Directive({ ... })
export class MyDirective {
  constructor(
    private viewContainerRef: ViewContainerRef,
    private componentFactoryResolver: ComponentFactoryResolver
  ) {}

  createMyComponent() {
    const componentFactory = this.componentFactoryResolver.resolveComponentFactory(MyComponent);
    this.viewContainerRef.createComponent(componentFactory);
  }
}
```

New version:

```
@Directive({ ... })
export class MyDirective {
  constructor(private viewContainerRef: ViewContainerRef) {}

  createMyComponent() {
    this.viewContainerRef.createComponent(MyComponent);
  }
}
```

Note: this work is based on @mhevery's PR #32742.


## PR Type
What kind of change does this PR introduce?

- [x] Feature

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

AndrewKushnir committed:
feat(core): add support for Types in ViewContainerRef.createComponent

With Ivy it's possible to create a `ComponentFactory` instance based on the generated Component def, thus the `ViewContainerRef.createComponent` can accept a Component instance and create a factory inside, rather than requiring additional complexity of creating a ComponentFactory outside (by invoking `ComponentFactoryResolver`).

This should simplify the API that is used for creating components dynamically.

DEPRECATION:

Angular no longer requires component factories to dynamically create components. The factory-based signature of the `ViewContainerRef.createComponent` function is deprecated in favor of a different signature that allows passing component classes instead.

AndrewKushnir commented on packages/core/test/bundling/todo/bundle.golden_symbols.json:
```diff
@@ -5,6 +5,12 @@
   {
     "name": "ChangeDetectionStrategy"
   },
+  {
```
Note: additional symbols in this test app are retained because of the extra code in the `ViewContainerRef` class (to create `ComponentFactory` based on a provided Component instance). The `ViewContainerRef` is used in the [NgFor](https://github.com/angular/angular/blob/master/packages/common/src/directives/ng_for_of.ts#L184) directive, **but** the `createComponent` is not actually used there (but the code still retains since it's a part of the `ViewContainerRef` class).

This can be resolved by introducing an internal version of the `ViewContainerRef` class (like `InternalViewContainerRef`) which would not contain the `createComponent` function. The `ViewContainerRef` then extend the `InternalViewContainerRef` and bring the `createComponent` function in. The `InternalViewContainerRef` might also be later used in other places like NgIf and NgSwitch directives (which don't use `ViewContainerRef.createComponent` calls).

IgorMinar replied:
fyi: this is unlikely to be worth the effort + complexity, but it's a good option to have for the future.

alfaproject commented:
Did you consider a method outside the class?

pkozlowski-opensource commented on packages/core/test/acceptance/view_container_ref_spec.ts:
```diff
@@ -1495,6 +1495,38 @@ describe('ViewContainerRef', () => {
          expect(fixture.debugElement.nativeElement.innerHTML).toContain('Child Component');
        });
+
+    it('should be able to create a component using Type', () => {
```
Could you please add more tests for "failure" paths? More specifically I would like to see what happens when we pass a type that is not a component.

pkozlowski-opensource replied:
What I've meant here is a test that exercises a code path where we pass a type that doesn't represent a component (ex. a directive or just a regular class / function). Not sure if we should have explicit checks for this but it would be interesting to see how it "explodes"

AndrewKushnir replied:
Thanks for the comment, I've added an extra check to the code (to make sure that we have `cmp` field on a type) and supporting tests for failure scenarios (+ tests for the `options` argument as well).

mary-poppins commented:
You can preview 51f9cae at https://pr43022-51f9cae.ngbuilds.io/.

mary-poppins commented:
You can preview 7ba448c at https://pr43022-7ba448c.ngbuilds.io/.

mary-poppins commented:
You can preview 841cd08 at https://pr43022-841cd08.ngbuilds.io/.

mary-poppins commented:
You can preview 4f5784e at https://pr43022-4f5784e.ngbuilds.io/.

AndrewKushnir commented:
@IgorMinar @pkozlowski-opensource @alxhub FYI I've updated the code to use a new signature for the `ViewContainerRef.createComponent` function. I've implemented it in a way that the `options` (2nd argument) is only available for the new signature (i.e. when component is provided) and the old API (with positional args) is available **only** when an old signature is used (i.e. when factory is provided). I did this intentionally, so that we deprecate the old API (and eventually remove it) and the new API only supports the new way of providing options (so it's an additional incentive to switch to the new signature).

Note: I also did a minor code reshuffling to avoid new import graph cycles. I ended up accepting 2 new cycles, since it'd be hard to get rid of them without moving more code around, which I can do separately as a followup.

I'd like to ask if you could take a look and provide an early feedback while I complete the work on remaining items:

- add tests to verify that old and new APIs work as expected
- deprecate the old signature of the `ViewContainerRef.createComponent` fn
- add more docs

Thank you.

mary-poppins commented:
You can preview aefaa91 at https://pr43022-aefaa91.ngbuilds.io/.

IgorMinar commented on packages/core/src/linker/view_container_ref.ts:
```diff
@@ -111,19 +115,40 @@ export abstract class ViewContainerRef {
+   * @param options An object that contains extra parameters:
+   *  * index: the index at which to insert the new component's host view into this container.
+   *           If not specified, appends the new view as the last entry.
+   *  * injector: the injector to use as the parent for the new component.
```
oh, and we should also document the relationship between this injector and the injector retrieved from the provided ngModuleRef.

pkozlowski-opensource replied:
Yeh, we need to find a good way of wording those injectors. In a nutshell a dynamically instantiated component will consult _both_ the `injector` and the one from `ngModuleRef` (in this order). In this sense it is hard to talk about one "parent" injector as there are two...

IgorMinar replied:
@pkozlowski-opensource you are right.

This is so confusing, and finding the code that is responsible for this is so tricky. I finally found it here: https://cs.opensource.google/angular/angular/+/master:packages/core/src/render3/component_ref.ts;l=138-145;drc=b5ab7aff433a67cddaa55e621d17b1a1b07b57c2

so yes, we should find a good wording here.

AndrewKushnir replied:
I think we should consider improving the `ViewContainerRef.createComponent` API as a whole. I've created a ticket #43602 to capture this. Aside from improving the wording, it'd be great to come up with some good examples too to better illustrate the arguments and their relation. I've improved some descriptions in this PR (original ones came from the docs of the existing `createComponent` function) and I'd propose to do more improvements as a part of #43602 (since we can incrementally add docs, but need to ship this change in v13).

IgorMinar commented on packages/core/src/linker/view_container_ref.ts:
```diff
@@ -111,19 +115,40 @@ export abstract class ViewContainerRef {
+   *  * ngModuleRef: an instance of the NgModuleRef that represent an NgModule.
+   *                 This information is used to retrieve corresponding NgModule injector.
+   *
+   * @returns The new component instance, containing the host view.
```
Don't you want to say that it returns an instance of a ComponentRef which contains the component instance and the host view?

IgorMinar reviewed (commented):
submitting initial api review feedback. I think we are on the right track here, let's clean it up a bit more and it's going to be great. Thank you, @AndrewKushnir!

pkozlowski-opensource commented on packages/core/src/linker/view_container_ref.ts:
```diff
@@ -111,19 +115,40 @@ export abstract class ViewContainerRef {
+   *  * index: the index at which to insert the new component's host view into this container.
+   *           If not specified, appends the new view as the last entry.
+   *  * injector: the injector to use as the parent for the new component.
+   *  * projectableNodes: list of DOM nodes that should be projected.
```
Actually it is a "list of lists" (or rather - an array of arrays - I don't think we should be using the term "list" here). The first dimension corresponds to the `<ng-content>` tag (there can be several) while the second dimension is the actual array of native nodes).

AndrewKushnir replied:
I've updated the doc as proposed (with the ng-content link), but I don't think it'd make sense to go into further details without examples. I can work on it as a followup (don't want to block this PR on this).

pkozlowski-opensource commented on packages/core/src/linker/view_container_ref.ts:
```diff
@@ -239,10 +270,68 @@ const R3ViewContainerRef = class ViewContainerRef extends VE_ViewContainerRef {
+      index = options.index;
+      injector = options.injector;
+      projectableNodes = options.projectableNodes;
+      ngModuleRef = options.ngModuleRef;
```
Would using `Object.extends` make it any better?

AndrewKushnir replied:
Not sure I understand the proposal... Could you please elaborate more?

pkozlowski-opensource replied:
Never mind, I was reading the code wrong, it is all good here.

pkozlowski-opensource commented on packages/core/src/linker/view_container_ref.ts:
```diff
@@ -184,6 +209,12 @@ export function injectViewContainerRef(): ViewContainerRef {
+// Helper method to check if provided value is a ComponentFactory and a Type (Component instance).
+function isComponentFactory(value: ComponentFactory<any>|
+                            Type<any>): value is ComponentFactory<any> {
+  return value && typeof value !== 'function';
```
How about we use the [existing](https://github.com/angular/angular/blob/fe691935091aaf7090864c8111a15f7cc7e53b6c/packages/core/src/interface/type.ts#L21-L23) `isType(...)` and reverse the check?

AndrewKushnir commented on goldens/size-tracking/aio-payloads.json:
```diff
@@ -15,7 +15,7 @@
       "uncompressed": {
         "runtime-es2017": 4841,
-        "main-es2017": 462493,
+        "main-es2017": 463023,
```
Current size at HEAD: 462867, actual size in this PR: 463023, so the delta is ~150 bytes, which is expected due to extra code that handles config options in the `ViewContainerRef.createComponent` fn.

AndrewKushnir commented on packages/core/src/linker/view_container_ref.ts:
```diff
@@ -111,19 +115,40 @@ export abstract class ViewContainerRef {
+   *           If not specified, appends the new view as the last entry.
+   *  * injector: the injector to use as the parent for the new component.
+   *  * projectableNodes: list of DOM nodes that should be projected.
+   *  * ngModuleRef: an instance of the NgModuleRef that represent an NgModule.
```
I've updated the docs as proposed.

Regarding the format: here is how it looks like right now:

https://pr43022-aefaa91.ngbuilds.io/api/core/ViewContainerRef#createComponent

I think the current layout is good, but we can always improve it as a followup.

AndrewKushnir commented on goldens/public-api/core/core.md:
```diff
@@ -1355,7 +1355,13 @@ export interface ViewChildrenDecorator {
 export abstract class ViewContainerRef {
     abstract clear(): void;
-    abstract createComponent<C>(componentFactory: ComponentFactory<C>, index?: number, injector?: Injector, projectableNodes?: any[][], ngModule?: NgModuleRef<any>): ComponentRef<C>;
+    abstract createComponent<C>(componentType: Type<C>, options?: {
```
Good point, thanks Igor. I'd like to keep this as is for now to preserve API consistency and I've added an item to the FW sync agenda. I agree that we should take care of this as a separate project and I think that'd be a great opportunity for community contributions.

AndrewKushnir commented on packages/core/src/linker/view_container_ref.ts:
```diff
@@ -240,9 +250,12 @@ const R3ViewContainerRef = class ViewContainerRef extends VE_ViewContainerRef {
 
   override createComponent<C>(
-      componentFactory: ComponentFactory<C>, index?: number|undefined,
+      componentFactoryOrType: ComponentFactory<C>|Type<C>, index?: number|undefined,
```
Yeah, while this is an ergonomic improvement, we can still do better. However the proposed change will require more design/discussions and we'd risk delaying this API and not making it in v13 :( With the `options` as an object we can improve things more after v13 without triggering a breaking change or loosing ergonomics.

mary-poppins commented:
You can preview 4de8e2d at https://pr43022-4de8e2d.ngbuilds.io/.

pkozlowski-opensource commented on packages/core/test/acceptance/view_container_ref_spec.ts:
```diff
@@ -1495,6 +1495,122 @@ describe('ViewContainerRef', () => {
+              viewContainerRef.createComponent(ChildA);
+
+              const projectableNode = document.createElement('div');
+              projectableNode.innerHTML = '[Projectable Node]';
```
nit: `innerHTML` -> [textContent](https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent)? I know that this is only test code, just my brain always "lights up" when I see `innerHTML` :-)

pkozlowski-opensource reviewed (commented):
LGTM overall. Still unsure about the size increase.

mary-poppins commented:
You can preview 206af28 at https://pr43022-206af28.ngbuilds.io/.

AndrewKushnir commented:
[Exploratory presubmit](http://test/OCL:399338127:BASE:399338144:1632796801389:ae12757f).

IgorMinar reviewed (approved):
LGTM! Thanks!

Reviewed-for: public-api, docs-packaging-and-releasing

mary-poppins commented:
You can preview 4774092 at https://pr43022-4774092.ngbuilds.io/.

AndrewKushnir commented:
[Presubmit](http://test/OCL:399582896:BASE:399582916:1632881555745:55670956).

mary-poppins commented:
You can preview 5c5aacf at https://pr43022-5c5aacf.ngbuilds.io/.

jessicajaniuk reviewed (approved):
LGTM :cookie:

reviewed-for: public-api, size-tracking

atscott reviewed (approved):
reviewed-for: public-api
reviewed-for: size-tracking

alxhub reviewed (approved):
Reviewed-for: public-api

AndrewKushnir commented:
Merge-assistance: we should be ok to proceed with merging, the `ci/angular: size` indicates that the todo app size got increased by over 2KB (which is a threshold for that check). The reason for the increase is that the `ViewContainerRef.createComponent` now refers to the `ComponentFactory` symbol, which retains other symbols used inside of it. For that minimal application, [experimental Ivy APIs](https://github.com/angular/angular/blob/master/packages/core/test/bundling/todo/index.ts#L202) are used for bootstrap. In real app the `ComponentFactory` would already be present, so no bundle size increase is expected. Moving forward, once we start reshaping/simplifying APIs based on Ivy capabilities, we'd be able to perform more refactoring and get size benefits (by avoiding deprecated codepaths).
